research/fed-sm/job_configs/fedsm_prostate/app/custom/learners/supervised_monai_prostate_learner.py

```diff
@@ -0,0 +1,183 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+
+import torch
+import torch.optim as optim
+from learners.supervised_learner import SupervisedLearner
+from monai.data import CacheDataset, DataLoader, Dataset, load_decathlon_datalist
+from monai.inferers import SimpleInferer
+from monai.losses import DiceLoss
+from monai.metrics import DiceMetric
+from monai.networks.nets.unet import UNet
+from monai.transforms import (
+    Activations,
+    AsDiscrete,
+    AsDiscreted,
+    Compose,
+    EnsureChannelFirstd,
+    EnsureType,
+    EnsureTyped,
+    LoadImaged,
+    Resized,
+    ScaleIntensityRanged,
+)
+from utils.custom_client_datalist_json_path import custom_client_datalist_json_path
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.app_constant import AppConstants
+
+
+class SupervisedMonaiProstateLearner(SupervisedLearner):
+    def __init__(
+        self,
+        train_config_filename,
+        aggregation_epochs: int = 1,
+        train_task_name: str = AppConstants.TASK_TRAIN,
+    ):
+        """MONAI Learner for prostate segmentation task.
+        It inherits from SupervisedLearner.
+
+        Args:
+            train_config_filename: path for config_3 file, this is an addition term for config_3 loading
+            aggregation_epochs: the number of training epochs for a round.
+            train_task_name: name of the task to train the model.
+
+        Returns:
+            a Shareable with the updated local model after running `execute()`
+        """
+        super().__init__(
+            aggregation_epochs=aggregation_epochs,
+            train_task_name=train_task_name,
+        )
+        self.train_config_filename = train_config_filename
+        self.config_info = None
+
+    def train_config(self, fl_ctx: FLContext):
+        """MONAI traning configuration
+        Here, we use a json to specify the needed parameters
+        """
+
+        # Load training configurations json
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_config_dir = ws.get_app_config_dir(fl_ctx.get_job_id())
+        train_config_file_path = os.path.join(app_config_dir, self.train_config_filename)
+        if not os.path.isfile(train_config_file_path):
+            self.log_error(
+                fl_ctx,
+                f"Training configuration file does not exist at {train_config_file_path}",
+            )
+        with open(train_config_file_path) as file:
+            self.config_info = json.load(file)
+
+        # Get the config_info
+        self.lr = self.config_info["learning_rate"]
```

Comment:
Ideally this and other attributes of the class should be initialized as None in `__init__()`.

Reply:
added